lib/server.js

```diff
@@ -0,0 +1,193 @@
+   * @public
+   */
+  listen(...args) {
+    return http.createServer(this.handle.bind(this)).listen(...args)
+  }
+
+  /**
+   * The main request handler handles all incoming requests.
+   * Integrate with your own server by calling this method and routing all requests to it.
+   * @param {http.IncomingMessage} req
+   * @param {http.ServerResponse} res
+   * @public
+   */
+  async handle(req, res) {
+    if (req.method !== 'GET' && req.method !== 'HEAD') {
+      return this.write(res, 405, 'Method not allowed')
+    }
+
+    const paths = url.parse(req.url || '')?.path?.split('/')
+
+    if (!paths || paths.length < 3) {
+      return this.write(res, 404, 'Malformed request')
+    }
+
+    try {
+      const [, receivedDigest, hex] = paths
+      var decodedUrl = this.verifyHmac(receivedDigest, hex)
+    } catch (err) {
+      return this.write(res, 403, 'Bad signature')
+    }
+
+    try {
+      var validUrl = await SafeHttpClient.checkUrl(decodedUrl)
+    } catch (err) {
+      const exception = /** @type {Error} */ (err)
+      return this.write(res, 400, exception.message)
+    }
+
+    const controller = new AbortController()
+    const {signal} = controller
+
+    req.on('close', () => {
+      controller.abort()
+    })
+
+    try {
+      // TODO: respect forwarded headers (check if not private IP)
+      const filterRequestHeaders = filterHeaders(defaultRequestHeaders)
+      const filterResponseHeaders = filterHeaders(defaultResponseHeaders)
+      const client = new SafeHttpClient(this.options.maxSize)
+      const {buffer, headers: resHeaders} = await client.safeFetch(validUrl, {
+        // @ts-expect-error: `IncomingHttpHeaders` can be passed to `Headers`
+        headers: filterRequestHeaders(new Headers(req.headers)),
+        method: req.method,
+        // We can't blindly follow redirects as the initial checkUrl
+        // might have been safe, but the redirect location might not be.
+        // SafeHttpClient will check the redirect location before following it.
+        redirect: 'manual',
+        signal
+      })
+
+      const headers = {
+        ...securityHeaders,
+        ...filterResponseHeaders(resHeaders),
+        Via: this.options.serverName
+      }
+
+      if (req.method === 'HEAD') {
+        res.writeHead(204, headers)
+        return res.end()
+      }
+      res.writeHead(200, headers)
+      res.write(buffer)
+      return res.end()
+    } catch (err) {
+      if (err instanceof HttpError) {
+        return this.write(res, err.statusCode, err.message)
+      }
+      if (err instanceof Error) {
+        if (err.name === 'AbortError') {
+          return
+        }
+        const msg = err.message || 'Internal server error'
+        return this.write(res, 500, msg)
+      }
+    }
+  }
+
+  /**
+   * @param {string} receivedDigest
+   * @param {string} hex
+   * @private
+   */
+  verifyHmac(receivedDigest, hex) {
+    // Hex-decode the URL
+    const decodedUrl = Buffer.from(hex, 'hex').toString()
```

Comment:
Nit: This is consistent with code below.

```suggestion
    const decodedUrl = String(Buffer.from(hex, 'hex'))
```